member_function_pointer_kernel_1.h does not compile with MSVC /vms

## Steps to Reproduce

* **Version**: 19.15
* **Where did you get dlib**: http://dlib.net/files/dlib-19.15.zip
* **Platform**: Windows 10 x64
* **Compiler**: MSVC 2015/2017, LLVM Clang-CL 8.0


```cpp
#include <dlib/config_reader.h>
int main() { }
```

Compile with

    cl main.cpp /I dlib-19.15 /vmg /vms /c

<details><summary>Output</summary>

```
main.cpp
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\xlocale(341): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(75): error C2287: 'dlib::mfp_kernel_1_base_class<0>::dummy': inheritance representation: 'single_inheritance' is less general than the required 'virtual_inheritance'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(120): note: see declaration of 'dlib::mfp_kernel_1_base_class<0>::dummy'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(80): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<0>::mp_null<dlib::mfp_kernel_1_base_class<0>::dummy>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(186): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<0>::mp_impl_T<dlib::mfp_kernel_1_base_class<0>::mp_null<dlib::mfp_kernel_1_base_class<0>::dummy>>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(205): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<0>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(75): error C2287: 'dlib::mfp_kernel_1_base_class<1>::dummy': inheritance representation: 'single_inheritance' is less general than the required 'virtual_inheritance'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(120): note: see declaration of 'dlib::mfp_kernel_1_base_class<1>::dummy'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(80): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<1>::mp_null<dlib::mfp_kernel_1_base_class<1>::dummy>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(186): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<1>::mp_impl_T<dlib::mfp_kernel_1_base_class<1>::mp_null<dlib::mfp_kernel_1_base_class<1>::dummy>>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(263): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<1>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(313): note: see reference to class template instantiation 'dlib::member_function_pointer<PARAM1,void,void,void>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(75): error C2287: 'dlib::mfp_kernel_1_base_class<2>::dummy': inheritance representation: 'single_inheritance' is less general than the required 'virtual_inheritance'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(120): note: see declaration of 'dlib::mfp_kernel_1_base_class<2>::dummy'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(80): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<2>::mp_null<dlib::mfp_kernel_1_base_class<2>::dummy>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(186): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<2>::mp_impl_T<dlib::mfp_kernel_1_base_class<2>::mp_null<dlib::mfp_kernel_1_base_class<2>::dummy>>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(322): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<2>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(371): note: see reference to class template instantiation 'dlib::member_function_pointer<PARAM1,PARAM2,void,void>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(75): error C2287: 'dlib::mfp_kernel_1_base_class<3>::dummy': inheritance representation: 'single_inheritance' is less general than the required 'virtual_inheritance'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(120): note: see declaration of 'dlib::mfp_kernel_1_base_class<3>::dummy'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(80): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<3>::mp_null<dlib::mfp_kernel_1_base_class<3>::dummy>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(186): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<3>::mp_impl_T<dlib::mfp_kernel_1_base_class<3>::mp_null<dlib::mfp_kernel_1_base_class<3>::dummy>>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(381): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<3>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(430): note: see reference to class template instantiation 'dlib::member_function_pointer<PARAM1,PARAM2,PARAM3,void>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(75): error C2287: 'dlib::mfp_kernel_1_base_class<4>::dummy': inheritance representation: 'single_inheritance' is less general than the required 'virtual_inheritance'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(120): note: see declaration of 'dlib::mfp_kernel_1_base_class<4>::dummy'
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(80): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<4>::mp_null<dlib::mfp_kernel_1_base_class<4>::dummy>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(186): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<4>::mp_impl_T<dlib::mfp_kernel_1_base_class<4>::mp_null<dlib::mfp_kernel_1_base_class<4>::dummy>>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(441): note: see reference to class template instantiation 'dlib::mfp_kernel_1_base_class<4>' being compiled
c:\users\niklas\desktop\dlib-19.15\dlib\member_function_pointer/member_function_pointer_kernel_1.h(491): note: see reference to class template instantiation 'dlib::member_function_pointer<PARAM1,PARAM2,PARAM3,PARAM4>' being compiled
```

</details>
<br/>

LLVM Clang-CL does not have a problem with this. Just substitute `cl` above with `clang-cl`.

I suppose MSVC may be correct here due to /vms instructing to use `single_inheritance`..? Unfortunately I must use `/vms` in this project. Could there be a way to overcome this issue with MSVC (eg. by explicitly stating the inheritance model for the classes in question)?

--- a/dlib/member_function_pointer/member_function_pointer_kernel_1.h
+++ b/dlib/member_function_pointer/member_function_pointer_kernel_1.h
@@ -116,8 +116,19 @@ namespace dlib
             }
         };
 
+        // MSVC with the /vms option, we get C2287 since the dummy class requires virtual
+        // inheritance. Adding the __virtual_inheritance specifier explicitly fixes the issue,
+        // but then Clang-CL no longer accepts it.
+        #if defined(_MSC_VER) && !defined(__clang__)
+            #define DLIB_MSVC_INHERITANCE_VIRTUAL __virtual_inheritance
+        #else
+            #define DLIB_MSVC_INHERITANCE_VIRTUAL
+        #endif
+
         struct dummy_base { virtual void nonnull() {}; virtual ~dummy_base(){}; int a; };
-        struct dummy : virtual public dummy_base{ void nonnull() {}; };
+        struct DLIB_MSVC_INHERITANCE_VIRTUAL dummy : virtual public dummy_base{ void nonnull() {}; };
+
+        #undef DLIB_MSVC_INHERITANCE_VIRTUAL
 
         typedef mp_impl_T<mp_null<dummy> > mp_null_impl;
     public:

